refactoring/extractfunc.go

```diff
@@ -123,15 +123,26 @@ func (r *ExtractFunc) checkForBlockStmt(node ast.Node, path []ast.Node) {
 }
 
 // If the users start selection is in the midway of a node, then the entire node is selected as the starting node
+// TODO reduce the calls of this, store a more useful version of it inside of our *ExtractFunc
 func (r *ExtractFunc) findStartPosition(node *ast.BlockStmt) ast.Stmt {
-	var stmt ast.Stmt = nil
+	var stmt ast.Stmt
+	line := r.Program.Fset.Position(r.SelectionStart).Line
 	for _, s := range node.List {
 		ast.Inspect(s, func(n ast.Node) bool {
 			if a, ok := n.(ast.Stmt); ok {
 				if r.SelectionStart == a.Pos() {
 					stmt = a
 					return false
-				} else if r.SelectionStart >= a.Pos() && r.SelectionStart < a.End() {
+				} else if r.Program.Fset.Position(a.Pos()).Line == line && // if it's on same line
+					r.SelectionStart < a.End() && // and less than end
+					(r.SelectionStart > a.Pos() || // and we're either in the middle of a node
+						stmt == nil || // or we haven't found anything
+						a.Pos()-r.SelectionStart < stmt.Pos()-r.SelectionStart) { // or closer than what we have found
+					// NOTE: [I] do not trust this approximation very much. it attempts to fudge the beginning of the
+					// selection to the right, b/c in the case that they selected the beginning of a line, we want to find
+					// the first statement, and skews left in the case that they made a selection in the middle of the node.
+					// the only case of invalid input I can think of is selecting _only_ whitespace, which should be
+					// guarded against by checking the end boundary coupled with there not being an ast.Stmt there. bueller?
```

Comment:
Ummmmmm... idunno?  All the tests still pass :)

I'm going to merge this in and then look into the selection code a bit more this afternoon.  I'll submit a pull request if I make more changes...